torch/multiprocessing/reductions.py

```diff
@@ -22,6 +22,9 @@ class StorageRef(object):
     def __init__(self, ptr):
         self.cdata = ptr
 
+    def __del__(self):
```

Comment:
In the above comment, I don't think anything sets "cdata" to None anymore.

If I understand correctly:

- `shared_cache` has a weak reference to a `StorageRef`
- `StorageRef` morally holds a weak reference to a THStorage pointer
- `THStorage` holds a strong reference to `StorageRef` via finalizer

So when all the THStorage strong refs go away:

- The finalizer runs (does nothing)
- The finalizer gets destructed (decrements StorageRef's ref count)
- StorageRef drops a weak reference of the THStorage
- This may lead to 0 weak refs on THStorage which frees the THStorage struct

It would be nice to simplify this, but I don't see an easy way to do it. It would be nice if a weak ref to a `torch.Storage` but a weakref to a `torch.Storage` Python object is not the same as a weak ref to a `THStorage`. (You could make it so by making `THStorage` own it's wrapper).

Reply:
Your understanding is right.

> You could make it so by making THStorage own it's wrapper

Right, and it would be very similar to what we do for Tensor today (to preserve `id()` across FFI crossings). I didn't do this, mostly because I saw a way to do it with the finalizer instead. Note that you are always going to need a finalizer, because the destructor for THStorage isn't going to know how to decref a Python object, so someone else has to give it a hand. If you want me to do it this way, I can; it's not too much more work.